File: Solution/src/com/solution/main/ReadInput.java
package com.solution.main;

import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

import com.solution.model.Tree;

public class ReadInput {

	public static Scanner getScanner(String filename) {
		InputStream is = ReadInput.class.getClassLoader().getResourceAsStream(filename);
		return new Scanner(is);
	}

	public static Tree getTree(String filename) {
		List<Integer> arr = new ArrayList<Integer>();

		Scanner sc = getScanner(filename);

		sc.forEachRemaining(t -> {
			arr.add(Integer.parseInt(t));
		});

		Tree t = doGetTreeInput(arr, 0, arr.size());

		return t;

	}

	private static Tree doGetTreeInput(List<Integer> arr, int low, int high) {
		if (( high-low ) == 0) {
			return null;
		}
		int r_index = (low + high) / 2;

		Tree t = new Tree(arr.get(r_index));

		Tree left = doGetTreeInput(arr, low, r_index);
		Tree right = doGetTreeInput(arr, r_index+1, high);

		t.setLeft(left);
		t.setRight(right);

		return t;

	}
}

File: src/main/kotlin/com/cssolutions/main/ReadInput.kt
package com.cssolutions.main

import com.cssolutions.model.Tree
import java.util.*

object ReadInput {
    fun getScanner(filename: String?): Scanner {
        val `is` = ReadInput::class.java.classLoader.getResourceAsStream(filename)
        val sc = Scanner(`is`)
        sc.useDelimiter("\n")
        return sc
    }

    @JvmStatic
	fun getTree(filename: String?): Tree? {
        val arr: MutableList<Int> = ArrayList()
        val sc = getScanner(filename)
        sc.useDelimiter(" ")
        sc.forEachRemaining { t: String -> arr.add(t.toInt()) }
        return doGetTreeInput(arr, 0, arr.size)
    }

    fun getIntArray(size: Int, arr: String): IntArray {
        val array = arr.split(" ".toRegex()).toTypedArray()
        val result = IntArray(size)
        for (i in 0 until size) {
            result[i] = array[i].toInt()
        }
        return result
    }

    private fun doGetTreeInput(arr: List<Int>, low: Int, high: Int): Tree? {
        if (high - low == 0) {
            return null
        }
        val r_index = (low + high) / 2
        val t = Tree(arr[r_index])
        val left = doGetTreeInput(arr, low, r_index)
        val right = doGetTreeInput(arr, r_index + 1, high)
        t.left = left
        t.right = right
        return t
    }
}